### PR TITLE
Add copy button to HighlightedCode component

### PR DESCRIPTION
**Pull request by michaeltmiller**
Closes https://github.com/code-chronicles-code/leetcode-curriculum/issues/152

I think it matches the overall vibe of the frontend, but I'm definitely open to feedback! For instance I'd be happy to add some simple icons or account for older browsers where [navigator.clipboard](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard#browser_compatibility) isn't supported.

### CODE REVIEW

**Review by miorel — COMMENTED** [review]
Yay, thanks so much for the PR!

And thanks for saying that the frontend has a vibe 😅 
It's definitely a part of the code that hasn't gotten any investment for a while, and I'd like to give it more attention in the near future.

Icons and other UI improvements would be very welcome, though don't feel obliged to do it here -- this PR already looks really good, and I think we can merge it with the current functionality. I just added a few comments regarding the implementation.

Regarding browser support, it seems like clipboard has been supported for a while, so I'm cool with assuming we have it. Though perhaps users with aggressive security settings could turn it off? I'm not sure how common that is. We could make a note to try checking for [permission](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API) and not render buttons if we don't have permission, but that's probably something for another PR as well.

Also, just wanted to confirm, are you alright with your contributions being distributed under the MIT license?

Thanks again! 🙏

File: workspaces/adventure-pack/css/style.css [hunk 1]
```diff
@@ -21,3 +21,25 @@ body,
...
+
+.highlighted-code {
+  position: relative;
+}
+
+.copy-button {
+  position: absolute;
+  top: 5px;
+  right: 5px;
+  z-index: 100;
```
[the remark this answers is not in the document]
Comment (michaeltmiller) [reply workspaces/adventure-pack/css/style.css:33]: Oops, sorry for committing this slop! This was an artifact of some earlier testing. Will address.

File: workspaces/adventure-pack/src/app/components/CopyButton.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import React, { useCallback, useState } from "react";
+
+type CopyButtonProps = {
+  textToCopy: string | string[];
+  style?: React.CSSProperties;
```
[the remark this answers is not in the document]
Comment (michaeltmiller) [reply workspaces/adventure-pack/src/app/components/CopyButton.tsx:5]: It is not used so you're right it should not exist.

File: workspaces/adventure-pack/src/app/components/CopyButton.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  className = "copy-button",
+}: CopyButtonProps) {
+  const [copied, setCopied] = useState(false);
+
+  const handleCopy = useCallback(() => {
+    navigator.clipboard
+      .writeText(textToCopy as string)
+      .then(() => {
+        setCopied(true);
+        setTimeout(() => setCopied(false), 2000);
```
[the remark this answers is not in the document]
Comment (michaeltmiller) [reply workspaces/adventure-pack/src/app/components/CopyButton.tsx:20]: Yeah I considered a few scenarios when deciding how robust this needed to be. The behavior you described is certainly not a polished experience. I'll leave a comment drawing attention to it. 

Also, the code as written isn't cleaning up the hanging `setTimeout` in the event that the component is unmounted before the callback handed `setTimeout` runs. I don't think recent versions of react still warn about setting state on an unmounted component, but still probably not a best practice.

File: workspaces/adventure-pack/src/app/components/CopyButton.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+export function CopyButton({
+  textToCopy,
+  className = "copy-button",
+}: CopyButtonProps) {
+  const [copied, setCopied] = useState(false);
+
+  const handleCopy = useCallback(() => {
+    navigator.clipboard
+      .writeText(textToCopy as string)
```
[the remark this answers is not in the document]
Comment (michaeltmiller) [reply workspaces/adventure-pack/src/app/components/CopyButton.tsx:17]: Yeah, will address. I'm quite new to TS so appreciate this feedback.

File: workspaces/adventure-pack/css/style.css [hunk 5]
```diff
@@ -21,3 +21,24 @@ body,
...
+.copy-button {
+  position: absolute;
+  top: 5px;
+  right: 5px;
+  background: rgba(255, 255, 255, 0.7);
+  border: none;
+  border-radius: 4px;
+  padding: 4px 8px;
+  cursor: pointer;
+  opacity: 0;
```
Comment (miorel) [inline workspaces/adventure-pack/css/style.css:38]: Random note, I'd be curious to understand the accessibility implications of 0 opacity, I believe some screen readers may treat it the same as visibility hidden: https://stackoverflow.com/a/43206381

We may also want to have opacity 1 when the button is focused, even if the container is not hovered.

But I think that's for a future investigation.

File: workspaces/adventure-pack/src/app/components/CopyButton.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+export function CopyButton({ textToCopy }: CopyButtonProps) {
+  const [copied, setCopied] = useState(false);
+
+  const handleCopy = useCallback(async () => {
+    try {
+      await navigator.clipboard.writeText(textToCopy);
+      setCopied(true);
+      // TODO: consider edge cases such as user clicks again within 2s or component is unmounted with pending setTimeout
```
Comment (miorel) [inline workspaces/adventure-pack/src/app/components/CopyButton.tsx:14]: One possible way to handle both of these is with an effect.

```ts
useEffect(() => {
  if (copied) {
    const timeoutID = setTimeout(() => setCopied(false), 2000);
    return () => clearTimeout(timeoutID);
  }
  return undefined;
}, [copied]);
```

**Review by miorel — APPROVED** [review]

**Comment by miorel** [comment]
Left a couple of comments for future consideration, but this looks good to merge. Thanks again!

Let me know if you want to take on anything else, I'd be happy to discuss some ideas 😄

**Pull request merged**